src/iter/blocks.rs

```diff
@@ -0,0 +1,90 @@
+use super::plumbing::*;
+use super::*;
+
+/// `ByBlocks` is a parallel iterator that consumes itself as a sequence
+/// of parallel blocks.
+///
+/// This struct is created by the [`by_blocks()`] method on [`IndexedParallelIterator`]
+///
+/// [`by_blocks()`]: trait.IndexedParallelIterator.html#method.by_blocks
+/// [`IndexedParallelIterator`]: trait.IndexedParallelIterator.html
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+#[derive(Debug, Clone)]
+pub struct ByBlocks<I, S> {
+    base: I,
+    sizes: S,
+}
+
+impl<I, S> ByBlocks<I, S>
+where
+    I: IndexedParallelIterator,
+    S: IntoIterator<Item = usize>,
+{
+    /// Creates a new `ByBlocks` iterator.
+    pub(super) fn new(base: I, sizes: S) -> Self {
+        ByBlocks { base, sizes }
+    }
+}
+
+struct BlocksCallback<S, C> {
+    sizes: S,
+    consumer: C,
+    len: usize,
+}
+
+impl<T, S, C> ProducerCallback<T> for BlocksCallback<S, C>
+where
+    C: UnindexedConsumer<T>,
+    S: Iterator<Item = usize>,
+{
+    type Output = C::Result;
+    fn callback<P: Producer<Item = T>>(mut self, mut producer: P) -> Self::Output {
+        let mut remaining_len = self.len;
+        let mut consumer = self.consumer;
+        // we need a local variable for the accumulated results
+        // we call the reducer's identity by splitting at 0
+        let (left_consumer, right_consumer, _) = consumer.split_at(0);
+        let mut leftmost_res = left_consumer.into_folder().complete();
+        consumer = right_consumer;
+
+        // now we loop on each block size
+        while remaining_len > 0 && !consumer.full() {
+            // we compute the next block's size
+            let size = self.sizes.next().unwrap_or(std::usize::MAX);
```

Comment:
There's a design choice here when `sizes.next()` returns `None`, with at least three options:

- Consume the remainder as one entire block, which is what you have now.
- Run the remainder for side effects, without feeding it into our consumer. `Skip` does this with its skipped prefix to match the semantics of `std::iter::Skip`.
- Ignore the remainder altogether, basically acting like the suffix of `Take`.

Did you consider this? I think behavior like `Take` might actually be a nice choice.

Another question is whether `size == 0` should be allowed. At the very least, it would be a wasted split here, which could just be considered the user's poor choice. Or we could make that impossible by having `S` produce `NonZeroUsize` items, but that makes it more annoying to write the iterator.

Reply:
Hmm. It seems quite surprising to me to ignore some of the elements. I would definitely not expect that. Ultimately, I think we want to offer users a choice of how to manage this -- another behavior I could imagine would be "go on using the last block size that got returned until we are complete".

I am wondering if we can express these semantics with some composed operators? I'd sort of like to say something like `.by_blocks().skip_remainder()` or something. We could just add inherent methods to the return type to allow users to change back and forth between those modes, no?

Reply:
It seems to me that the user can express most of those semantics in the iterator itself:

- current MAX behavior: `my_iter.chain(iter::once(usize::MAX))`
- repeat behavior: `my_iter.chain(iter::repeat(last))`
- my suggested `Take` behavior: just `my_iter`, and let its `None` be the end
- run remainder without consuming: ???

> We could just add inherent methods to the return type to allow users to change back and forth between those modes, no?

Yeah, that's possible too.

Reply:
another might be asserting that there is no remainder

Reply:
I think at minimum we should document these techniques, and i might prefer a convenient way to express them

Reply:
if we only allow constant sizes and doubling sizes in the public interface then this choice about what to do when the size iterator ends is of private concern.
i used to have both versions, one like an implicit take when it runs out and the other one forcing to consume everything. so there is indeed something to choose here.

i would favor consuming it all since I happened to get some bugs missing some elements.

Reply:
also, i have some questions like:
- should i add unit tests ? where ?
- should i add benches ? where ?

Reply:
I would be in favor of only allowing the two options until we know there is demand for something more general. As for unit tests, we typically just add `#[test]` in the modules, I think? Alternatively, you can add tests into the `tests` directory but we don't usually do that. (Is this right, @cuviper?)

For benchmarks, we generally modify the rayon-bench project.

Reply:
For tests, there are a few generic sanity checks to add: `tests/clones.rs`, `tests/debug.rs`, and `src/compile_fail/must_use.rs`. Beyond that, adding unit `#[test]` directly in the module is fine, or break into a `tests` submodule if there are a lot. Doc-test examples are also nice to serve both doc and testing roles.